metrics-exporter-prometheus/src/formatting.rs

```diff
@@ -121,24 +114,58 @@ pub fn write_metric_line<T, T2>(
     buffer.push('\n');
 }
 
-fn add_unit(buffer: &mut String, unit: &str) {
-    const SUM_SUFFIX: &str = "_sum";
-    const COUNT_SUFFIX: &str = "_count";
-    const BUCKET_SUFFIX: &str = "_bucket";
-
-    if buffer.ends_with(SUM_SUFFIX) {
-        let suffix_pos = buffer.len() - SUM_SUFFIX.len();
-        buffer.insert(suffix_pos, '_');
-        buffer.insert_str(suffix_pos + 1, unit);
-    } else if buffer.ends_with(COUNT_SUFFIX) {
-        let suffix_pos = buffer.len() - COUNT_SUFFIX.len();
-        buffer.insert(suffix_pos, '_');
-        buffer.insert_str(suffix_pos + 1, unit);
-    } else if buffer.ends_with(BUCKET_SUFFIX) {
-        let suffix_pos = buffer.len() - BUCKET_SUFFIX.len();
-        buffer.insert(suffix_pos, '_');
-        buffer.insert_str(suffix_pos + 1, unit);
-    } else {
+fn add_metric_name(
+    buffer: &mut String,
+    name: &str,
+    unit: Option<Unit>,
+    suffix: Option<&'static str>,
+) {
+    buffer.push_str(name);
+    if let Some(unit) = unit {
+        add_unit_if_missing(buffer, unit);
+    }
+    if let Some(suffix) = suffix {
+        add_suffix_if_missing(buffer, suffix);
+    }
+}
+
+/// Adds a suffix to the metric name if it is not already in the name.
+fn add_suffix_if_missing(buffer: &mut String, suffix: &str) {
+    if !buffer.ends_with(suffix) {
+        buffer.push('_');
+        buffer.push_str(suffix);
+    }
+}
+
+/// Adds a unit to the metric name if it is not already in the name.
+/// If the metric ends with a known suffix, we try to insert the unit before the suffix.
+/// Otherwise, we append the unit to the end of the metric name.
+fn add_unit_if_missing(buffer: &mut String, unit: Unit) {
+    const KNOWN_SUFFIXES: [&str; 4] = ["_sum", "_count", "_bucket", "_total"];
+
+    let unit = match unit {
+        Unit::Count => {
+            // For count, we don't suffix the unit.
+            return;
+        }
+        Unit::Percent => "ratio",
```

Comment:
@tobz I'm not sure where the "percent" -> "ratio" conversion was there originally.